File: src/main/java/com/vjiki/music/service/MessageServiceImpl.java
package com.vjiki.music.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import com.vjiki.music.dto.CreateMessageRequest;
import com.vjiki.music.dto.MessageReactionRequest;
import com.vjiki.music.dto.MessageReactionResponse;
import com.vjiki.music.dto.MessageResponse;
import com.vjiki.music.entity.Message;
import com.vjiki.music.mapper.MessageMapper;
import com.vjiki.music.pagination.CreatedAtIdCursorCodec;
import com.vjiki.music.pagination.CursorPageResponse;
import com.vjiki.music.repository.ChatParticipantRepository;
import com.vjiki.music.repository.ChatRepository;
import com.vjiki.music.repository.MessageReactionRepository;
import com.vjiki.music.repository.MessageReadRepository;
import com.vjiki.music.repository.MessageRepository;
import com.vjiki.music.repository.UserRepository;

@Service
public class MessageServiceImpl implements MessageService {

    private final MessageRepository messageRepository;
    private final MessageReactionRepository messageReactionRepository;
    private final MessageReadRepository messageReadRepository;
    private final ChatRepository chatRepository;
    private final ChatParticipantRepository chatParticipantRepository;
    private final UserRepository userRepository;

    public MessageServiceImpl(MessageRepository messageRepository,
                              MessageReactionRepository messageReactionRepository,
                              MessageReadRepository messageReadRepository,
                              ChatRepository chatRepository,
                              ChatParticipantRepository chatParticipantRepository,
                              UserRepository userRepository) {
        this.messageRepository = messageRepository;
        this.messageReactionRepository = messageReactionRepository;
        this.messageReadRepository = messageReadRepository;
        this.chatRepository = chatRepository;
        this.chatParticipantRepository = chatParticipantRepository;
        this.userRepository = userRepository;
    }

    @Override
    public List<MessageResponse> getMessagesByChatId(UUID chatId, UUID userId1, UUID userId2) {
        return messageRepository.findMessagesByChatIdAndUsers(chatId, userId1, userId2).stream()
                .map(MessageMapper::toResponse)
                .toList();
    }

    @Override
    public CursorPageResponse<MessageResponse> getMessagesPage(UUID chatId, int limit, String cursor) {
        int safeLimit = Math.min(Math.max(limit, 1), 100);
        Pageable pageable = PageRequest.of(0, safeLimit + 1);

        CreatedAtIdCursorCodec.Cursor decoded = CreatedAtIdCursorCodec.decodeOrBadRequest(cursor);
        List<Message> messages = decoded == null
                ? messageRepository.findMessagesPageFirst(chatId, pageable)
                : messageRepository.findMessagesPageAfter(chatId, decoded.createdAt(), decoded.id(), pageable);

        if (messages.isEmpty()) {
            return new CursorPageResponse<>(List.of(), null, false);
        }

        boolean hasNext = messages.size() > safeLimit;
        List<Message> slice = hasNext ? messages.subList(0, safeLimit) : messages;

        // Native queries don't load relationships, so re-read the slice through JPA to get senders.
        List<UUID> messageIds = slice.stream().map(Message::getId).toList();
        Map<UUID, Message> messagesWithSenders = new HashMap<>();
        for (Message message : messageRepository.findAllById(messageIds)) {
            messagesWithSenders.put(message.getId(), message);
        }

        List<MessageResponse> items = new ArrayList<>(slice.size());
        for (Message message : slice) {
            Message withSender = messagesWithSenders.getOrDefault(message.getId(), message);
            items.add(MessageMapper.toResponse(withSender));
        }

        Message last = slice.getLast();
        String nextCursor = (hasNext && last.getCreatedAt() != null)
                ? CreatedAtIdCursorCodec.encode(last.getCreatedAt(), last.getId())
                : null;

        return new CursorPageResponse<>(items, nextCursor, hasNext);
    }

    @Override
    @Transactional
    public MessageResponse createMessage(CreateMessageRequest request) {
        if (!chatRepository.existsById(request.getChatId())) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Chat not found");
        }

        if (chatParticipantRepository.findByChatIdAndUserId(request.getChatId(), request.getSenderId()).isEmpty()) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "User is not a participant in this chat");
        }

        if (!userRepository.existsById(request.getSenderId())) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Sender not found");
        }

        if (request.getReplyToId() != null) {
            Message replyToMessage = messageRepository.findById(request.getReplyToId())
                    .orElseThrow(() -> new ResponseStatusException(
                            HttpStatus.NOT_FOUND, "Reply-to message not found"));

            if (!replyToMessage.getChatId().equals(request.getChatId())) {
                throw new ResponseStatusException(
                        HttpStatus.BAD_REQUEST, "Reply-to message does not belong to this chat");
            }
        }

        UUID messageId = UUID.randomUUID();

        messageRepository.insertMessage(
                messageId,
                request.getChatId(),
                request.getSenderId(),
                request.getReplyToId(),
                request.getMessageType(),
                request.getContent(),
                request.getSongId(),
                request.getAttachmentCount() == null ? 0 : request.getAttachmentCount());

        chatRepository.updateChatTimestamp(request.getChatId());

        Message messageWithSender = messageRepository.findByIdWithSender(messageId)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.INTERNAL_SERVER_ERROR, "Failed to retrieve saved message"));

        return MessageMapper.toResponse(messageWithSender);
    }

    @Override
    @Transactional
    public void deleteMessage(UUID messageId) {
        Message message = messageRepository.findById(messageId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found"));

        if (Boolean.TRUE.equals(message.getIsDeleted())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Message is already deleted");
        }

        messageRepository.deleteMessage(messageId);
    }

    @Override
    public List<MessageReactionResponse> getMessageReactions(UUID messageId) {
        if (!messageRepository.existsById(messageId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found");
        }

        return messageReactionRepository.findByMessageId(messageId).stream()
                .map(reaction -> MessageReactionResponse.builder()
                        .messageId(reaction.getMessageId())
                        .userId(reaction.getUserId())
                        .emoji(reaction.getEmoji())
                        .createdAt(reaction.getCreatedAt())
                        .build())
                .toList();
    }

    @Override
    @Transactional
    public void addReaction(MessageReactionRequest request) {
        if (!messageRepository.existsById(request.getMessageId())) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found");
        }

        if (!userRepository.existsById(request.getUserId())) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found");
        }

        messageReactionRepository.insertReaction(request.getMessageId(), request.getUserId(), request.getEmoji());
    }

    @Override
    @Transactional
    public void removeReaction(UUID messageId, UUID userId, String emoji) {
        if (!messageRepository.existsById(messageId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found");
        }

        if (!userRepository.existsById(userId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found");
        }

        messageReactionRepository.deleteReaction(messageId, userId, emoji);
    }

    @Override
    @Transactional
    public void markAsRead(UUID messageId, UUID userId) {
        Message message = messageRepository.findById(messageId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found"));

        if (!userRepository.existsById(userId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found");
        }

        if (chatParticipantRepository.findByChatIdAndUserId(message.getChatId(), userId).isEmpty()) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "User is not a participant in this chat");
        }

        messageReadRepository.markAsRead(messageId, userId);

        chatParticipantRepository.updateLastReadMessageId(message.getChatId(), userId, messageId);
    }
}

File: bin/main/com/vjiki/music/service/MessageServiceImpl.kt
package com.vjiki.music.service

import com.vjiki.music.dto.CreateMessageRequest
import com.vjiki.music.dto.MessageReactionRequest
import com.vjiki.music.dto.MessageReactionResponse
import com.vjiki.music.dto.MessageResponse
import com.vjiki.music.mapper.MessageMapper.toResponse
import com.vjiki.music.pagination.CreatedAtIdCursorCodec
import com.vjiki.music.pagination.CursorPageResponse
import com.vjiki.music.repository.ChatParticipantRepository
import com.vjiki.music.repository.ChatRepository
import com.vjiki.music.repository.MessageReadRepository
import com.vjiki.music.repository.MessageReactionRepository
import com.vjiki.music.repository.MessageRepository
import com.vjiki.music.repository.UserRepository
import org.springframework.data.domain.PageRequest
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException
import java.util.*

@Service
class MessageServiceImpl(
    private val messageRepository: MessageRepository,
    private val messageReactionRepository: MessageReactionRepository,
    private val messageReadRepository: MessageReadRepository,
    private val chatRepository: ChatRepository,
    private val chatParticipantRepository: ChatParticipantRepository,
    private val userRepository: UserRepository
) : MessageService {

    override fun getMessagesByChatId(chatId: UUID, userId1: UUID, userId2: UUID): List<MessageResponse> {
        return messageRepository.findMessagesByChatIdAndUsers(chatId, userId1, userId2)
            .map { it.toResponse() }
    }

    override fun getMessagesPage(chatId: UUID, limit: Int, cursor: String?): CursorPageResponse<MessageResponse> {
        val safeLimit = limit.coerceIn(1, 100)
        val pageable = PageRequest.of(0, safeLimit + 1)

        val decoded = CreatedAtIdCursorCodec.decodeOrBadRequest(cursor)
        val messages = if (decoded == null) {
            messageRepository.findMessagesPageFirst(chatId, pageable)
        } else {
            messageRepository.findMessagesPageAfter(
                chatId = chatId,
                cursorCreatedAt = decoded.createdAt,
                cursorId = decoded.id,
                pageable = pageable
            )
        }

        if (messages.isEmpty()) {
            return CursorPageResponse(
                items = emptyList(),
                nextCursor = null,
                hasNext = false
            )
        }

        val hasNext = messages.size > safeLimit
        val slice = if (hasNext) messages.take(safeLimit) else messages

        // Fetch senders for messages (native queries don't load relationships)
        val messageIds = slice.map { it.id }
        val messagesWithSenders = if (messageIds.isNotEmpty()) {
            messageRepository.findAllById(messageIds).associateBy { it.id }
        } else {
            emptyMap()
        }

        val items = slice.map { message ->
            val messageWithSender = messagesWithSenders[message.id] ?: message
            messageWithSender.toResponse()
        }

        val last = slice.lastOrNull()
        val nextCursor = if (hasNext && last?.createdAt != null) {
            CreatedAtIdCursorCodec.encode(last.createdAt, last.id)
        } else {
            null
        }

        return CursorPageResponse(
            items = items,
            nextCursor = nextCursor,
            hasNext = hasNext
        )
    }

    @Transactional
    override fun createMessage(request: CreateMessageRequest): MessageResponse {
        // Validate chat exists
        if (!chatRepository.existsById(request.chatId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Chat not found")
        }

        // Validate sender is participant
        val participant = chatParticipantRepository.findByChatIdAndUserId(request.chatId, request.senderId)
        if (participant.isEmpty) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN, "User is not a participant in this chat")
        }

        // Validate sender exists
        if (!userRepository.existsById(request.senderId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Sender not found")
        }

        // Validate replyTo message exists and belongs to same chat
        if (request.replyToId != null) {
            val replyToMessage = messageRepository.findById(request.replyToId)
                .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Reply-to message not found") }
            
            if (replyToMessage.chatId != request.chatId) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Reply-to message does not belong to this chat")
            }
        }

        // Generate UUID for the new message
        val messageId = UUID.randomUUID()

        // Insert message using native query
        messageRepository.insertMessage(
            id = messageId,
            chatId = request.chatId,
            senderId = request.senderId,
            replyToId = request.replyToId,
            messageType = request.messageType,
            content = request.content,
            songId = request.songId,
            attachmentCount = request.attachmentCount
        )

        // Update chat's updated_at timestamp
        chatRepository.updateChatTimestamp(request.chatId)

        // Fetch with sender relation for response
        val messageWithSender = messageRepository.findByIdWithSender(messageId)
            .orElseThrow { ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to retrieve saved message") }

        return messageWithSender.toResponse()
    }

    @Transactional
    override fun deleteMessage(messageId: UUID) {
        // Validate message exists
        val message = messageRepository.findById(messageId)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found") }

        if (message.isDeleted) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Message is already deleted")
        }

        // Soft delete the message
        messageRepository.deleteMessage(messageId)
    }

    override fun getMessageReactions(messageId: UUID): List<MessageReactionResponse> {
        // Validate message exists
        if (!messageRepository.existsById(messageId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found")
        }

        val reactions = messageReactionRepository.findByMessageId(messageId)
        return reactions.map { reaction ->
            MessageReactionResponse(
                messageId = reaction.messageId,
                userId = reaction.userId,
                emoji = reaction.emoji,
                createdAt = reaction.createdAt
            )
        }
    }

    @Transactional
    override fun addReaction(request: MessageReactionRequest) {
        // Validate message exists
        if (!messageRepository.existsById(request.messageId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found")
        }

        // Validate user exists
        if (!userRepository.existsById(request.userId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "User not found")
        }

        // Insert reaction using native query
        messageReactionRepository.insertReaction(
            messageId = request.messageId,
            userId = request.userId,
            emoji = request.emoji
        )
    }

    @Transactional
    override fun removeReaction(messageId: UUID, userId: UUID, emoji: String) {
        // Validate message exists
        if (!messageRepository.existsById(messageId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found")
        }

        // Validate user exists
        if (!userRepository.existsById(userId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "User not found")
        }

        // Delete reaction using native query
        messageReactionRepository.deleteReaction(
            messageId = messageId,
            userId = userId,
            emoji = emoji
        )
    }

    @Transactional
    override fun markAsRead(messageId: UUID, userId: UUID) {
        // Validate message exists
        val message = messageRepository.findById(messageId)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Message not found") }

        // Validate user exists
        if (!userRepository.existsById(userId)) {
            throw ResponseStatusException(HttpStatus.NOT_FOUND, "User not found")
        }

        // Validate user is participant in the chat
        val participant = chatParticipantRepository.findByChatIdAndUserId(message.chatId, userId)
        if (participant.isEmpty) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN, "User is not a participant in this chat")
        }

        // Mark as read using native query
        messageReadRepository.markAsRead(messageId, userId)

        // Update participant's last_read_message_id
        chatParticipantRepository.updateLastReadMessageId(message.chatId, userId, messageId)
    }
}

